tactics/tactics.ml

```diff
@@ -3038,8 +3050,13 @@ let specialize (c,lbind) ipat =
         | Context.Rel.Declaration.LocalAssum _::lp' , t::l' ->
           let sigma,hd' = rebuild_lambdas sigma lp' (args@[t]) hd l' in
           sigma,hd'
+        | Context.Rel.Declaration.LocalDef _::lp' , _ ->
+           (* letins are have been reduced in l and should anyway not
+              correspond to an arg, we ignore them. *)
+          let sigma,hd' = rebuild_lambdas sigma lp' args hd l in
+          sigma,hd'
```

Comment:
So this means we are not reabstracting over the let-ins right? This is probably fine.

Reply:
letins dont need to be abstracted (they do not correspond to an application to the hypothesis).
Moreover the line
```
let clause = make_clenv_binding env sigma (c,typ_of_c) lbind in
```
removes the letins from the clause anyway. So the letins are there in lprod but not in l.

Reply:
My point was that your code behaves as if you had performed a ζ reduction on the type of the term being specialized, i.e. let-ins in the middle of the telescope get expanded. Anyways I don't think it matters much.

Reply:
Yes and no.

- You are right the letins are reduced from the start by make_clenv_binding. I would rather avoid that and keep the letins in the type but I don't think we can build a clause and solve it without removing the letins. Can we?
- Even if we find a way to not reduce the letins they do not correspond to arguments of the hypothesis. So we would still ignore them when reabstracting stuff.